File: go/roman-numerals/roman_numerals.go
package romannumerals

import "errors"

// MIN_ROMAN is the minimum valid integer for Roman numeral conversion.
const MIN_ROMAN = 1

// MAX_ROMAN is the maximum valid integer for Roman numeral conversion.
const MAX_ROMAN = 3999

// ToRomanNumeral converts an integer to its Roman numeral representation.
// It returns an error if the input is outside the valid range of 1 to 3999.
func ToRomanNumeral(input int) (string, error) {
	if input < MIN_ROMAN || input > MAX_ROMAN {
		return "", errors.New("input out of range")
	}

	number := ""

	for current := input; current > 0; {
		atom, value := nextAtom(current)
		current -= value
		number += atom
	}

	return number, nil
}

// nextAtom finds the largest Roman numeral atom that can be subtracted from a number.
// It returns the Roman numeral string and its integer value.
func nextAtom(number int) (string, int) {
	for _, value := range atomsValues {
		if number >= value {
			return atomsFromValue[value], value
		}
	}
	return "", 0
}

// atomsValues is a sorted slice of integer values for Roman numeral atoms.
var atomsValues = []int{
	1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1,
}

// atomsFromValue maps integer values to their corresponding Roman numeral strings.
var atomsFromValue = map[int]string{
	1000: "M",
	900:  "CM",
	500:  "D",
	400:  "CD",
	100:  "C",
	90:   "XC",
	50:   "L",
	40:   "XL",
	10:   "X",
	9:    "IX",
	5:    "V",
	4:    "IV",
	1:    "I",
}

File: lua/roman-numerals/roman-numerals.lua
--- A module for converting integers to Roman numerals.
local roman <const> = {}

--- A constant lookup table containing Roman numeral symbols and their integer values.
--- The table is ordered from smallest to largest value, which is crucial for the conversion algorithm.
local ROMAN_RUNES <const> = {
  { 1,  "I" }, { 4, "IV" }, { 5, "V" }, { 9, "IX" },
  { 10, "X" }, { 40, "XL" }, { 50, "L" }, { 90, "XC" },
  { 100,  "C" }, { 400, "CD" }, { 500, "D" }, { 900, "CM" },
  { 1000, "M" }
}

--- Converts an integer to its Roman numeral representation.
--- The algorithm iterates through the Roman numeral values from largest to smallest,
--- subtracting the largest possible value from the number and collecting the corresponding symbol in a table.
--- Finally, the symbols are concatenated to form the result string.
--- @param n integer The integer to convert.
--- @return string -- The Roman numeral representation of the number.
function roman.to_roman(n)
  local current = n
  local result = {}

  for i = #ROMAN_RUNES, 1, -1 do
    local value = ROMAN_RUNES[i][1]
    local rune = ROMAN_RUNES[i][2]
    
    while current >= value do
      current = current - value
      table.insert(result, rune)
    end
  end

  return table.concat(result)
end

return roman